oci/pull.bzl

```diff
@@ -37,24 +37,164 @@ oci_image(
 """
 
 load("@aspect_bazel_lib//lib:paths.bzl", "BASH_RLOCATION_FUNCTION")
+load("@aspect_bazel_lib//lib:base64.bzl", "base64")
+load("@aspect_bazel_lib//lib:repo_utils.bzl", "repo_utils")
+
+def _strip_host(url):
+    # TODO: a principled way of doing this
+    return url.replace("http://", "").replace("https://", "").replace("/v1/", "")
+
+def _file_exists(rctx, path):
+    result = rctx.execute(["stat", path])
+    return result.return_code == 0
+
+def _get_auth_file_path(rctx):
+    # this is the standard path where registry credentials are stored
+    config_path = "{}/.docker/config.json".format(rctx.os.environ["HOME"])
+
+    # set config path to DOCKER_CONFIG env if present
+    if "DOCKER_CONFIG" in rctx.os.environ:
+        config_path = rctx.os.environ["DOCKER_CONFIG"]
+
+    if _file_exists(rctx, config_path):
+        return config_path
+
+    # https://docs.podman.io/en/latest/markdown/podman-login.1.html#authfile-path
+    XDG_RUNTIME_DIR = "{}/.config".format(rctx.os.environ["HOME"])
+    if "XDG_RUNTIME_DIR" in rctx.os.environ:
+        XDG_RUNTIME_DIR = rctx.os.environ["XDG_RUNTIME_DIR"]
+
+    config_path = "{}/containers/auth.json".format(XDG_RUNTIME_DIR)
+
+    if _file_exists(rctx, config_path):
+        return config_path
+
+    return None
+
+def _auth_anonymous(rctx, registry, repository, identifier):
+    """A function that performs anonymous auth for docker registry.
+
+    Args:
+        rctx: repository context
+        registry: registry url
+        repository: image repository
+        identifier: tag or digest
+
+    Returns:
+        A dict for rctx.download#auth
+    """
+    pattern = {}
+    if registry == "index.docker.io":
+        scope = "repository:{}:pull".format(repository)
+        rctx.download(
+            url = ["https://auth.docker.io/token?scope={}&service=registry.docker.io".format(scope)],
+            output = "auth_anonymous.json",
+        )
+        auth_raw = rctx.read("auth_anonymous.json")
+        auth = json.decode(auth_raw)
+        pattern = {
+            "type": "pattern",
+            "pattern": "Bearer <password>",
+            "password": auth["token"],
+        }
+
+    return pattern
+
+def _auth_basic(rctx, registry, repository, identifier):
```

Comment:
@gregmagolan do you think it's worth sharing code from our npm repository rule via something in bazel_lib?